### PR TITLE
feat(cli): Support nested_type in generator

### LINKED ISSUE #1397: awscc provider not working
Hi,

I have a small Java cdktf application and I am trying to add the awscc provider to the cdktf.json.
I tried several formats, with and without version:
  _"terraformProviders": [
    "awscc@~> 0.8"
  ],_

 and the cdktf get command is failing with the stack trace:
_Error: unknown type undefined
    at $8e.renderAttributeType (/home/mircea/.npm-global/lib/node_modules/cdktf-cli/bundle/bin/cdktf.js:43:50957)
    at $8e.renderAttributesForBlock (/home/mircea/.npm-global/lib/node_modules/cdktf-cli/bundle/bin/cdktf.js:43:51145)
    at $8e.resourceFrom (/home/mircea/.npm-global/lib/node_modules/cdktf-cli/bundle/bin/cdktf.js:43:49571)
    at QX.parse (/home/mircea/.npm-global/lib/node_modules/cdktf-cli/bundle/bin/cdktf.js:43:53591)
    at YX.emitProvider (/home/mircea/.npm-global/lib/node_modules/cdktf-cli/bundle/bin/cdktf.js:44:4652)
    at new YX (/home/mircea/.npm-global/lib/node_modules/cdktf-cli/bundle/bin/cdktf.js:44:4371)
    at aAe.generateTypeScript (/home/mircea/.npm-global/lib/node_modules/cdktf-cli/bundle/bin/cdktf.js:50:6162)
    at async aAe.generate (/home/mircea/.npm-global/lib/node_modules/cdktf-cli/bundle/bin/cdktf.js:50:6251)
    at async /home/mircea/.npm-global/lib/node_modules/cdktf-cli/bundle/bin/cdktf.js:370:6323
⠏ downloading and generating modules and providers...
unknown type undefined_

I know there is no prebuilt provider for awscc, but I was under the assumption that it's not required that a provider be prebuilt to use it in cdktf.

If there is of any importance, I am trying to use a recently added feature to a Lambda event source mapping, more details here:
https://github.com/hashicorp/terraform-provider-aws/issues/22068

Or is there something that I am doing wrong?

### PR DESCRIPTION
**Pull request by jsteinich**
Support for tfproto v6
Fixes #1397

### CODE REVIEW

**Comment by jsteinich** [comment]
This does break compatibility in that the new cli will generate binding not compatible with the old cdktf lib since the order of parameters was changed in `ComplexObject` constructor.
Technically some minor breaking changes as well. Mostly this helps the user by returning more information for nested computed lists. For nested computed maps it's maybe worse (`getStringMapAttribute` -> `StringMap.lookup`). We had already planned on making some changes there, so I may just follow this up with that work.

I've looked quite a bit at how this changes existing providers and also confirmed that providers using the new protocol successfully build. Could probably use some tests to confirm that synth works correctly with nested types.

File: packages/@cdktf/provider-generator/lib/get/__tests__/generator/__snapshots__/provider.test.ts.snap [hunk 1]
```diff
@@ -163,7 +163,7 @@ export interface AwsProviderAssumeRole {
   readonly sessionName?: string;
 }
 
-export function awsProviderAssumeRoleToTerraform(struct?: AwsProviderAssumeRole | cdktf.IResolvable): any {
+export function awsProviderAssumeRoleToTerraform(struct?: AwsProviderAssumeRole): any {
```
[the remark this answers is not in the document]
Comment (jsteinich) [reply packages/@cdktf/provider-generator/lib/get/__tests__/generator/__snapshots__/provider.test.ts.snap:166]: This wasn't an intentional change, but it didn't seem to break any tests so I hadn't looked into it. I'll do some more digging to see why it changed.

Adding `IResolvable` to these functions was done to make sure things that can't be represented by tokens natively wouldn't break.

File: packages/@cdktf/provider-generator/lib/get/generator/emitter/struct-emitter.ts [hunk 2]
```diff
@@ -185,7 +185,29 @@ export class StructEmitter {
...
+    } else if (struct.nestingMode === "map") {
+      this.code.line(
+        `* @param complexObjectKey the key of this item in the map`
+      );
+      this.code.line(`*/`);
+      this.code.openBlock(
+        `public constructor(terraformResource: cdktf.IInterpolatingParent, terraformAttribute: string, complexObjectKey: string)`
+      );
+      this.code.line(
+        `super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectKey);`
```
[the remark this answers is not in the document]
Comment (jsteinich) [reply packages/@cdktf/provider-generator/lib/get/generator/emitter/struct-emitter.ts:209]: Good catch. I'll fix.

File: packages/@cdktf/provider-generator/lib/get/__tests__/generator/__snapshots__/provider.test.ts.snap (hunk 1 above)
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/@cdktf/provider-generator/lib/get/__tests__/generator/__snapshots__/provider.test.ts.snap:166]: I think it'll probably be fine :)

**Review by ansgarm — APPROVED** [review]

File: packages/@cdktf/provider-generator/lib/get/__tests__/generator/__snapshots__/provider.test.ts.snap (hunk 1 above)
[the remark this answers is not in the document]
Comment (jsteinich) [reply packages/@cdktf/provider-generator/lib/get/__tests__/generator/__snapshots__/provider.test.ts.snap:166]: There was a missing case for single item sets in resource-parser. I had forgotten adding that in.
The `IResolvable` is only needed for a [specific case](https://github.com/hashicorp/terraform-cdk/blob/f97577c4b91833771665be3b0672492cf9bccbea/packages/%40cdktf/provider-generator/lib/get/generator/models/attribute-type-model.ts#L110), so removing this is completely fine.

**Review by ansgarm — CHANGES_REQUESTED** [review]
I just tested this with the `awscc` provider and there seems to be an issue with invalid TypeScript code created for datasources:
```
// throttle - computed: true, optional: false, required: false
  private _throttle = new { [key: string]: DataAwsccApigatewayUsagePlanApiStagesThrottle } | cdktf.IResolvableMap(this, "throttle");
  public get throttle() {
    return this._throttle;
  }
```
(`"hashicorp/awscc@~> 0.16.0"`; `.gen/providers/awscc/data-awscc-apigateway-usage-plan.ts`)

The TS errors are:
> This expression is not constructable.
>  Type '{ [x: number]: any; DataAwsccApigatewayUsagePlanApiStagesThrottle: any; }' has no construct signatures.

> Property 'IResolvableMap' does not exist on type 'typeof import("/Users/ansgar/projects/playground/test-get-awscc/node_modules/cdktf/lib/index")'.

**Review by ansgarm — APPROVED** [review]
Really cool @jsteinich 👍

**Pull request merged**